grid-client/deployer/deployment_deployer.go

```diff
@@ -185,24 +208,189 @@ func (d *DeploymentDeployer) updateStateFromDeployments(ctx context.Context, dl
+			}
 		}
 	}
 
-	if contractID, ok := dl.NodeDeploymentID[dl.NodeID]; ok && contractID != 0 {
-		dl.ContractID = contractID
-		if !workloads.Contains(d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID], dl.ContractID) {
-			d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID] = append(d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID], dl.ContractID)
+	return nil
+}
+
+func (d *DeploymentDeployer) calculateNetworksUsedIPs(ctx context.Context, dls []*workloads.Deployment) (map[string]map[uint32][]byte, error) {
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	var errs error
+	usedHosts := make(map[string]map[uint32][]byte)
+
+	// calculate used host IDs per network
+	for _, dl := range dls {
+		mu.Lock()
+		if _, ok := usedHosts[dl.NetworkName]; !ok {
+			usedHosts[dl.NetworkName] = make(map[uint32][]byte)
+		}
+
+		if _, ok := usedHosts[dl.NetworkName][dl.NodeID]; ok {
+			continue
+		}
+		mu.Unlock()
+
+		wg.Add(1)
+		go func(dl *workloads.Deployment) {
+			defer wg.Done()
+
+			nodeClient, err := d.tfPluginClient.NcPool.GetNodeClient(d.tfPluginClient.SubstrateConn, dl.NodeID)
+			if err != nil {
+				mu.Lock()
+				defer mu.Unlock()
+				errs = multierror.Append(errs, errors.Wrapf(err, "could not get node client for node %d", dl.NodeID))
+				return
+			}
+
+			privateIPs, err := nodeClient.NetworkListPrivateIPs(ctx, dl.NetworkName)
+			if err != nil {
+				mu.Lock()
+				defer mu.Unlock()
+				errs = multierror.Append(errs, errors.Wrapf(err, "could not list private ips from node %d", dl.NodeID))
+				return
+			}
+
+			network := d.tfPluginClient.State.Networks.GetNetwork(dl.NetworkName)
+			ipRange := network.GetNodeSubnet(dl.NodeID)
+
+			_, ipRangeCIDR, err := net.ParseCIDR(ipRange)
+			if err != nil {
+				mu.Lock()
+				defer mu.Unlock()
+				errs = multierror.Append(errs, errors.Wrapf(err, "invalid ip range %s", ipRange))
+				return
+			}
+
+			// get used host IDs of the deployment node using the private IPs list of the node
+			for _, privateIP := range privateIPs {
+				parsedPrivateIP := net.ParseIP(privateIP).To4()
+
+				if ipRangeCIDR.Contains(parsedPrivateIP) {
+					mu.Lock()
+					defer mu.Unlock()
+					usedHosts[dl.NetworkName][dl.NodeID] = append(usedHosts[dl.NetworkName][dl.NodeID], parsedPrivateIP[3])
+				}
+			}
+		}(dl)
+	}
+
+	wg.Wait()
+	return usedHosts, errs
+}
+
+func (d *DeploymentDeployer) assignPrivateIPs(ctx context.Context, dls []*workloads.Deployment) error {
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	var errs error
+
+	usedHosts, err := d.calculateNetworksUsedIPs(ctx, dls)
+	if err != nil {
+		return errors.Wrap(err, "couldn't calculate network used ips")
+	}
+
+	for _, dl := range dls {
+		if len(dl.Vms) == 0 {
+			continue
 		}
-		network := d.tfPluginClient.State.Networks.GetNetwork(dl.NetworkName)
-		updateNetworkUsedIPs(&network, dl)
-		d.tfPluginClient.State.Networks[dl.NetworkName] = network
+
+		wg.Add(1)
+		go func(dl *workloads.Deployment) {
```

Comment:
here also there is no blocking calls. I don't thing goroutines would perform better